### ISSUE TITLE
Validation error with spec that contains refs to local files

### ISSUE THREAD

**Issue by jbub** [go-swagger/go-swagger#729]
## Problem statement

When validating spec that includes schema from external file, it fails with:

```
$ swagger version
version: 0.7.4
commit: 6b8226461ff1d53e4b56ea0ea80d126aa3e9d7b5

$ swagger validate spec.yaml
The swagger spec at "spec.yaml" is invalid against swagger specification 2.0. see errors :
- invalid character 'y' in literal true (expecting 'r')
```

Swagger tools reports that the spec is valid:

```
$ swagger-tools --version
0.10.1

$ swagger-tools validate -v spec.yaml

Validation Details:

  Swagger Version: 2.0
  Swagger file: spec.yaml

Swagger document is valid
```

## Swagger specification

https://github.com/jbub/swaggerrepro/blob/master/spec.yaml

## Steps to reproduce

```
git clone https://github.com/jbub/swaggerrepro.git
swagger validate spec.yaml
```

**Comment by wittrock** [comment]
Has anyone figured out yet if there's a specific workaround for this in go-swagger, or is all external file reference functionality broken?

**Comment by casualjim** [comment]
at this moment it's just broken. Personally I'm strapped for time to make this work at this very moment, I'm hoping to spend some time on this during the week of christmas to fix it.
But if somebody wants to fix this sooner then I'd gladly merge

**Comment by danieljy** [comment]
@casualjim I'm finding that local references work (swagger 0.7.4, and swagger 0.8.0), but only if the references are JSON formatted files:

swagger.yaml:
```
      'conditions':
        type: array
        items:
          $ref: './swagger-test.yaml#definitions/Condition'
```
swagger-test.yaml
Works:
```
{
  "definitions": {
    "Condition": {
      "type": "object",
      "properties": {
        "name": {
          "type": "string"
        }
      }
    }
  }
}
```

Fails:
```
definitions:
  Condition:
    type: object
    properties:
      name:
        type: string
```

It would seem that a JSON parser is actually trying to resolve the references, as opposed to the YAML parser.

Any ideas?

**Comment by casualjim** [comment]
That is indeed the default behavior, but you can replace the loader function in your own program.

https://github.com/go-openapi/spec/blob/master/expander.go#L183

If you load your spec with the loads package this should be done for you:

https://github.com/go-openapi/loads/blob/master/spec.go#L54

if that isn't the case then you may need to enlist yaml as a recognized format:

https://github.com/go-swagger/go-swagger/blob/master/cmd/swagger/swagger.go#L27-L29

**Comment by danieljy** [comment]
@casualjim Before I go deeper here, it seems as though your examples suggest that go-swagger already is doing the right thing (the last link) - so why am I see the behavior when I run:

`swagger generate server -f swagger.yaml`

**Comment by casualjim** [comment]
can you try to build from master and see if that fixes your issue?

**Comment by casualjim** [comment]
on master there is also a new command called flatten, does it still occur after flattening the spec?

**Comment by danieljy** [comment]
I've got some updates:

Using code from master, I'm able to achieve the following:

```
../go-swagger validate swagger-local.yaml
The swagger spec at "swagger-local.yaml" is valid against swagger specification 2.0
```
But when I try to generate the server directly from this yaml file, I get a message like the following:

```
../go-swagger generate server -f swagger-local.yaml
2017/04/13 09:51:36 building a plan for generation
2017/04/13 09:51:36 planning definitions
object has no key "LatLng"
```

Here's an excerpt from swagger-local.yaml, demonstrating a reference

```
    properties:
      ...
      position:
        $ref: ./base.yaml#/definitions/LatLng

```

And here's an excerpt from base.yaml, which is in the same directory:

```
definitions:
  LatLng:
    type: object
    properties:
      lat:
        type: number
        format: double
      lng:
        type: number
        format: double
    required: ['lat', 'lng']
```

If i try flattening first, I get a different error, which comes from a reference from within the base.yaml file:

```
../go-swagger flatten swagger-local.yaml > swagger-local.json
2017/04/13 09:51:19 schema analysis [#/definitions/testLocation/properties/conditions]: object has no key "TestCondition"
make: *** [swagger-local.json] Error 1
```


```

TestCondition:
...

TestLocation:
    type: object
    properties:
      position:
        $ref: '#/definitions/LatLng'
      extent:
        $ref: '#/definitions/LatLngLine'
      info:
        $ref: '#/definitions/LocationMetadata'
      length:
        type: number
        format: double
      conditions:
        type: array
        items:
          $ref: '#/definitions/TestCondition'
    required: ['info']

...
```

So basically, each method has a problem.

In general, what's the right way to set these references up? Furthermore, my end goal is really to share types between different apis, so is there a best practice for that?

Thanks!

**Comment by casualjim** [comment]
I really want to see your entire spec or equivalent repro.
I suspect there might still be an issue in the spec somewhere and if not it's something that needs to be added to the tests so that it can remain solved.

**Comment by danieljy** [comment]
Okay here's a version that suffers from the same issues:

swagger-template.yaml:

```
swagger: '2.0'
info:
  title: Test
  version: "1"
host: test.com
basePath: /
produces:
  - application/json
schemes:
  - http
paths:
  /a/b:
    get:
      operationId: get_ab
      parameters:
        - name: pilot_area_only
          in: query
          type: boolean
          default: false
      responses:
        200:
          description: Test
          headers:
            Access-Control-Allow-Origin:
              type: string
              default: '*'
          schema:
            type: object
            properties:
              results:
                $ref: '#/definitions/LocalDefs'
        default:
          description: Unexpected error
          schema:
            $ref: '#/definitions/Error'
definitions:
  LocalDef:
    type: object
    properties:
      position:
        $ref: './base.yaml#/definitions/LatLng'
      extent:
        $ref: './base.yaml#/definitions/LatLngLine'
  LocalDefs:
    type: array
    items:
      $ref: '#/definitions/LocalDef'
  Error:
    type: object
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string
```

base.yaml:
```
definitions:
  LatLng:
    type: object
    properties:
      lat:
        type: number
        format: double
      lng:
        type: number
        format: double
    required: ['lat', 'lng']
  LatLngLine:
    type: array
    items:
      $ref: '#/definitions/LatLng'
```
As is, if I try to generate directly from swagger-template.yaml, I get the following error:

```
../go-swagger generate server -f swagger-local.yaml
2017/04/13 11:48:45 building a plan for generation
2017/04/13 11:48:45 planning definitions
object has no key "LatLng"
```

If instead I flatten first:
```
../go-swagger flatten swagger-local.yaml > swagg-flattened.json
danieljy:parking-availability danielyehuda$ ../go-swagger generate server -f swagg-flattened.json
The swagger spec at "swagg-flattened.json" is invalid against swagger specification 2.0. see errors :
- object has no key "LatLng"
```

If i change base.yaml to look like this:

```
definitions:
  LatLng:
    type: object
    properties:
      lat:
        type: number
        format: double
      lng:
        type: number
        format: double
    required: ['lat', 'lng']
  LatLngLine:
    type: array
    items:
      $ref: '#./base.yaml/definitions/LatLng'
```

I am able to fully generate a server, though it requires that I manually generate the models from base.yaml (they don't get generated as part of the `generate server` command on the main file) and then I have the problem that the generated operations files don't include the models.

If I try flattening first, it just hangs.


Hope this helps!

**Comment by danieljy** [comment]
I came upon: https://github.com/go-swagger/go-swagger/issues/402, which seems to be the same issue I'm seeing when I change my base.yaml to the form directly above.

It seems the issue is closed, but it's unclear to me where i should inserting the `x-go-package` and `x-go-name` to make the generated code properly import and reference the models.

Any ideas?

**Comment by casualjim** [comment]
You're being hit by 2 different bugs.

* when you don't flatten, the generator seems to not resolve the external file in generation.
* when you flatten, then the properties get a new casing for some reason, making the ref indeed invalid because it points to `#definitions/LatLng` instead of `#definitions/latLng`

I think the second bug is easier to solve than the first one

**Comment by danieljy** [comment]
@casualjim What about my most recent comment - seems I can actually get everything working without flattening, except for getting the generated code to reference separately generated models.

I'd actually prefer to have a single package with the generated shared models rather than flattening (which I believe would cause me to generate differently namespaced models depending on where I reference them).

In other words, is there a proper way of setting `x-go` params somewhere so that my generated code properly imports external models?

**Comment by casualjim** [comment]
I think this might work. I think only x-go-name was implemented in the bug you mentioned.

```yaml
swagger: '2.0'
info:
  title: Test
  version: "1"
host: test.com
basePath: /
produces:
  - application/json
schemes:
  - http
paths:
  /a/b:
    get:
      operationId: get_ab
      parameters:
        - name: pilot_area_only
          in: query
          type: boolean
          default: false
      responses:
        200:
          description: Test
          headers:
            Access-Control-Allow-Origin:
              type: string
              default: '*'
          schema:
            type: object
            properties:
              results:
                $ref: '#/definitions/LocalDefs'
        default:
          description: Unexpected error
          schema:
            $ref: '#/definitions/Error'
definitions:
  LatLng:
    type: object
    x-go-type:
      import:
        package: "github.com/user/package"
        alias: "base"
      type: "LatLng"
  LatLngLine:
    type: object
    x-go-type:
      import:
        package: "github.com/user/package"
        alias: "base"
      type: "LatLngLine"
  LocalDef:
    type: object
    properties:
      position:
        $ref: './base.yaml#/definitions/LatLng'
      extent:
        $ref: './base.yaml#/definitions/LatLngLine'
  LocalDefs:
    type: array
    items:
      $ref: '#/definitions/LocalDef'
  Error:
    type: object
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string
````
